bundles/org.openhab.core.addon.marketplace.karaf/src/main/java/org/openhab/core/addon/marketplace/karaf/internal/community/CommunityKarafAddonHandler.java

```diff
@@ -0,0 +1,189 @@
+import org.openhab.core.OpenHAB;
+import org.openhab.core.addon.Addon;
+import org.openhab.core.addon.marketplace.MarketplaceAddonHandler;
+import org.openhab.core.addon.marketplace.MarketplaceHandlerException;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A {@link CommunityKarafAddonHandler} implementation, which handles add-ons as kar files and installs them through
+ * the standard OSGi feature repository installation mechanism.
+ *
+ * @author Kai Kreuzer - Initial contribution and API
+ * @author Yannick Schaus - refactoring
+ * @author Jan N. Klug - refactor to support kar files
+ *
+ */
+@Component(immediate = true)
+@NonNullByDefault
+public class CommunityKarafAddonHandler implements MarketplaceAddonHandler {
+    private static final String KAR_CONTENT_TYPE = "application/vnd.openhab.feature;type=karfile";
+    private static final String BUNDLE_CACHE_PATH = OpenHAB.getUserDataFolder() + File.separator + "marketplace"
+            + File.separator + "kar";
+    private static final List<String> SUPPORTED_EXT_TYPES = List.of("binding");
+    private static final String KAR_DOWNLOAD_URL_PROPERTY = "kar_download_url";
+
+    private final Logger logger = LoggerFactory.getLogger(CommunityKarafAddonHandler.class);
+
+    private final FeaturesService featuresService;
+
+    @Activate
+    public CommunityKarafAddonHandler(@Reference FeaturesService featuresService) {
+        this.featuresService = featuresService;
+        ensureCachedBundlesAreInstalled();
+    }
+
+    @Override
+    public boolean supports(String type, String contentType) {
+        // we support only certain extension types, and only as pure OSGi bundles
+        return SUPPORTED_EXT_TYPES.contains(type) && KAR_CONTENT_TYPE.equals(contentType);
+    }
+
+    @Override
+    @SuppressWarnings("null")
+    public boolean isInstalled(String addonId) {
+        try {
+            File addonDirectory = getAddonCacheDirectory(addonId);
+            List<URI> repositories = Arrays.stream(featuresService.listRepositories()).map(Repository::getURI)
+                    .collect(Collectors.toList());
+            if (addonDirectory.exists() && addonDirectory.isDirectory()) {
+                return Files.list(addonDirectory.toPath()).filter(path -> path.endsWith(".kar")).findFirst()
+                        .map(Path::toUri).map(repositories::contains).orElse(false);
+            }
+        } catch (Exception e) {
+            logger.warn("Failed to determine installation status for {}: ", addonId, e);
+        }
+
+        return false;
+    }
+
+    @Override
+    public void install(Addon addon) throws MarketplaceHandlerException {
+        URL sourceUrl;
+        try {
+            sourceUrl = new URL((String) addon.getProperties().get(KAR_DOWNLOAD_URL_PROPERTY));
```

Comment:
```suggestion
        try {
            URL sourceUrl = new URL((String) addon.getProperties().get(KAR_DOWNLOAD_URL_PROPERTY));
```

Reply:
Fixed. Also changed in the handler where it was copied from.